File: src/lab3/UserType.java
package lab3;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;

public interface UserType<T> {
    public String typeName();
    public T create();
    public T clone(T o);
    public T readValue(InputStreamReader in) throws IOException;
    public T parseValue(String string);
    public Comparator<T> getTypeComparator();
}

File: src/lab5scala/UserType.scala
package lab5scala

import java.io.InputStreamReader
import java.util.Comparator

trait UserType[T] {
  def typeName: String
  def create: T
  def clone(o: T): T
  def readValue(in: InputStreamReader): T
  def parseValue(string: String): T
  def getTypeComparator: Comparator[T]
}
